src/internal/eval-tidy.c

```diff
@@ -160,14 +160,14 @@ sexp* rlang_new_data_mask(sexp* bottom, sexp* top) {
   sexp* data_mask;
 
   if (bottom == r_null) {
-    bottom = KEEP(r_new_environment(r_empty_env, 0));
+    bottom = KEEP(r_new_environment(r_empty_env, 100));
     data_mask = bottom;
   } else {
     check_data_mask_input(bottom, "bottom");
     // Create a child because we don't know what might be in `bottom`
     // and we need to clear its contents without deleting any object
     // created in the data mask environment
-    data_mask = KEEP(r_new_environment(bottom, 0));
+    data_mask = KEEP(r_new_environment(bottom, 100));
```

Comment:
The increase to 100 is a bit unexpected, I think.

So previously using `0` would result in the internal default of `29` being used. I thought that this was a little high, honestly.

Isn't it somewhat rare to add additional bindings to a data mask? With the exception of: `mask$.data <- new_data_pronoun(mask)`

So I would have set the size to `4 + buffer`, where `4` comes from the 4 items you poke into the environment below, and `buffer` is maybe 3 or 4 in case the user wants to add `$.data <- ` or a small number of other things to the mask

I only bring this up because using a much smaller default hash table size can improve performance when you have to create a large number of environments. This approximates the difference:

``` r
bench::mark(
  slow = {
    for (i in seq_len(1e5L)) {
      vector("list", 100L) # large hash table
    }
    
  },
  fast = {
    for (i in seq_len(1e5L)) {
      vector("list", 7L) # small hash table
    }
  }
)
#> Warning: Some expressions had a GC in every iteration; so filtering is disabled.
#> # A tibble: 2 x 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 slow         75.5ms   84.8ms      11.8    84.8MB     35.5
#> 2 fast         17.9ms   21.7ms      47.9    10.2KB     24.0
```

<sup>Created on 2020-10-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

Reply:
I was also thinking the environment created in `rlang_new_ctxt_pronoun()` could be much smaller. Maybe with a size of 2-3, since I doubt that ever gets modified? (at least that is what im guessing) https://github.com/r-lib/rlang/blob/4227d04a325763611564202a1ce95d52391e3da9/src/internal/eval-tidy.c#L54

Reply:
I can't reproduce your results:

```r
bench::mark(
  `0` = vector("list", 0),
  `30` = vector("list", 30),
  `100` = vector("list", 100),
  check = FALSE,
  iterations = 1e5
)[1:8]
#> # A tibble: 3 x 8
#>   expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>
#> 1 0             546ns    660ns  1376479.        0B    13.8  99999     1
#> 2 30            621ns    775ns  1084134.      288B    10.8  99999     1
#> 3 100           624ns    882ns   861974.      848B     8.62 99999     1
```

Isn't it weird that your slow run has 84MB of allocation while your fast run has 10KB?

In general, the difference between allocating a small and large vector is not that important. It's more about avoiding many different allocations.

Reply:
> Isn't it somewhat rare to add additional bindings to a data mask?

`mutate()` (which however uses as-data-mask rather than new-data-mask) would add bindings for new columns example, so it's probably not that rare. If you're using `new_data_mask()` you probably want to fill the mask with some bindings? In general there is no reason to use `eval_tidy()` unless you're masking data.

Anyway, given the small 100ns overhead that's probably a good enough default.

I do wonder where comes the difference between our benchmarks though, it seems like there is a bug somewhere.

Reply:
Two things I thought about regarding `mutate()`.

First, Romain's new PR actually was going to change to use `new_data_mask()` before you updated this one to initialize the mask with the expected size. I think he was essentially initializing the mask to the correct size himself.
https://github.com/tidyverse/dplyr/pull/5520/files#diff-d9709c495664bec3d8718ad24bf9db7391406f7ddac818694254efe79b2fe2acR74-R79

Second, regarding:

> `mutate()` would add bindings for new columns

I thought about this as well, and almost left a comment about it, but then I went and looked at the new implementation. _Technically_ it adds bindings into the enclosing environment of the data mask.

https://github.com/tidyverse/dplyr/pull/5520/files#diff-75499699eb24243ad9de81f1ea71d00d2cb1be968b57459fde2653a5006f5391R65

Whether this is correct or not, I'm not sure, but I think what Romain was trying to do is to install the new columns from `mutate()` into his original mask env that he created with the correct size. Since `new_data_mask()` creates a child env from `bottom`, he was inserting back into that original parent env.